### PR TITLE
Decouple buffer deallocation from ffi and allow creating buffers from rust vec

### LINKED ISSUE #1516: Allow creating buffers from externally owned memory like Vec or String
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

I'd like to process data that is coming from outside arrow using arrow compute kernels without copying the data.

The FFI api does not seem to support this use case at the moment. The only way I found to create an `FFI_ArrowArray` was with an already existing `ArrayData` struct. Creating `ArrayData` requires at least one `Buffer`, and creating a `Buffer` from foreign memory again needs an `FFI_ArrowArray`.

**Describe the solution you'd like**

The `Buffer` already supports two ways of deallocating memory.

 - Via arrows cache-aligned allocator
 - With a callback in an `FFI_ArrowArray`

The second option can be generalized so that the callback can be any type using a dyn trait.

**Describe alternatives you've considered**

An alternative could be to make it easier to create an `FFI_ArrowArray` from a single memory region, but that seems to be more complex and also less general.

**Additional context**

This functionality could also allow zero-copy conversion from arrow2.

### PR DESCRIPTION
**Pull request by jhorstmann**
# Which issue does this PR close?

Closes #1516

# Rationale for this change
 
 

I'd like to process data that is coming from outside arrow using arrow compute kernels without copying the data.

The FFI api does not seem to support this use case at the moment. The only way I found to create an `FFI_ArrowArray` was with an already existing `ArrayData` struct. Creating `ArrayData` requires at least one `Buffer`, and creating a `Buffer` from foreign memory again needs an `FFI_ArrowArray`.

# What changes are included in this PR?

The owner of an allocation is tracked via a trait object instead of an `FFI_ArrowArray`. This allows zero-copy interoperability between arrow buffers and most collection types from rust stdlib or other crates. The most common example would be `Vec` or `String`, and it should also be possible to access arrow2 buffers this way.

# Are there any user-facing changes?

The function `from_unowned` was marked as deprecated and delegates to a new function `from_custom_allocation`. The "unowned" part is abit misleading since the data is owned by the `FFI_ArrayArray`.

The other changes should be api compatible. The `bytes` module is `pub(crate)`, so moving the `Deallocation` enum should be fine.

### CODE REVIEW

File: arrow/src/alloc/mod.rs [hunk 1]
```diff
@@ -121,3 +124,30 @@ pub unsafe fn reallocate<T: NativeType>(
...
+    /// Native deallocation, using Rust deallocator with Arrow-specific memory alignment
+    Native(usize),
+    /// Foreign interface, via a callback
+    Foreign(Arc<dyn Allocation>),
+}
+
+impl Debug for Deallocation {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        match self {
+            Deallocation::Native(capacity) => {
```
Comment (alamb) [inline arrow/src/alloc/mod.rs:145]: This seems inconsistent with the comment above. This code says `capacity` but the comment above says `Native(usize)` contains alignment

**Review by alamb — COMMENTED** [review]
I think this is really cool -- nice work @jhorstmann 

In terms of naming, I would like to suggest we keep it backwards compatible for a while (so as not to have to bump major versions for a bit)

So that would mean
1. Keep the old interface named `from_unowned`, mark it deprecated, change impl to call `from_foreign` 
2. Add new `from_unowned()` that does what you suggest


cc @viirya and @sunchao

**Comment by alamb** [comment]
BTW I ran this under MIRI and it looks good:

```shell
cargo +nightly miri test -p arrow -- test_string_data_from_foreigntest_string_data_from_foreign
...
running 1 test
test array::data::tests::test_string_data_from_foreign ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 1072 filtered out

   Doc-tests arrow

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 145 filtered out; finished in 0.03s

```
🎉

File: arrow/src/alloc/mod.rs (hunk 1 above)
Comment (jhorstmann) [reply arrow/src/alloc/mod.rs:145]: I think the comment on `Native(usize)` means that the memory was allocated using arrow's allocator, which uses a specific alignment depending on platform. I'll update the comment to make that clearer.

File: arrow/src/alloc/mod.rs [hunk 2]
```diff
@@ -121,3 +124,30 @@ pub unsafe fn reallocate<T: NativeType>(
         handle_alloc_error(Layout::from_size_align_unchecked(new_size, ALIGNMENT))
     })
 }
+
+/// The owner of an allocation, that is not natively allocated.
+/// The trait implementation is responsible for dropping the allocations once no more references exist.
+pub trait Allocation: RefUnwindSafe {}
```
Comment (jhorstmann) [inline arrow/src/alloc/mod.rs:130]: The trait bound on `RefUnwindSafe` was needed to make one test compile. I'll need to check that in more detail and document why it is required.

Comment (viirya) [inline arrow/src/alloc/mod.rs:130]: nit: `ForeignAllocation`?

**Review by viirya — COMMENTED** [review]
This looks good idea to me. Thanks.

File: arrow/src/alloc/mod.rs (hunk 2 above)
Comment (jhorstmann) [reply arrow/src/alloc/mod.rs:130]: Other names I had considered were `Owner` or `AllocationOwner`. The `Native` / `Foreign` terminology no longer fits nicely, at least I consider a Rust `Vec` to be more "native" than our custom allocator.

Maybe we could rename the `Deallocation` variants to make this clearer. The enum was in a non-public module before (and should probably stay `pub(crate)`). I'm thinking of `Arrow(usize)` and `Custom(Arc<dyn Allocation>)`.

File: arrow/src/alloc/mod.rs [hunk 3]
```diff
@@ -121,3 +124,30 @@ pub unsafe fn reallocate<T: NativeType>(
...
+
+impl<T: RefUnwindSafe> Allocation for T {}
+
+/// Mode of deallocating memory regions
+pub enum Deallocation {
+    /// Native deallocation, using Rust deallocator with Arrow-specific memory alignment
+    Native(usize),
+    /// Foreign interface, via a callback
```
Comment (viirya) [inline arrow/src/alloc/mod.rs:138]:
```suggestion
    /// Foreign interface
```

I think callback is specified for `FFI_ArrowArray`.

File: arrow/src/alloc/mod.rs (hunk 2 above)
Comment (viirya) [reply arrow/src/alloc/mod.rs:130]: `Arrow(usize)` and `Custom(Arc<dyn Allocation>)` sounds good to me.

Comment (jhorstmann) [reply arrow/src/alloc/mod.rs:130]: I added an explicit test that `Buffer` is `UnwindSafe` to make this requirement clearer.

File: arrow/src/buffer/immutable.rs [hunk 4]
```diff
@@ -86,18 +84,41 @@ impl Buffer {
...
     /// * `ptr` - Pointer to raw parts
     /// * `len` - Length of raw parts in **bytes**
-    /// * `data` - An [ffi::FFI_ArrowArray] with the data
+    /// * `data` - An [crate::ffi::FFI_ArrowArray] with the data
     ///
     /// # Safety
     ///
     /// This function is unsafe as there is no guarantee that the given pointer is valid for `len`
     /// bytes and that the foreign deallocator frees the region.
+    #[deprecated(
```
Comment (alamb) [inline arrow/src/buffer/immutable.rs:93]: 👍

**Review by alamb — APPROVED** [review]
I think this looks great @jhorstmann  -- thank you. @viirya / @tustvold  / @sunchao any last thoughts before I merge it in?

I believe it is an API change, so we will bump the major version of arrow to use it (but that is probably ok given stuff like https://github.com/apache/arrow-rs/pull/1510 which will require a new version as well

**Comment by alamb** [comment]
There is a minor doctest failure: https://github.com/apache/arrow-rs/runs/5756218032?check_suite_focus=true

**Review by viirya — APPROVED** [review]
Looks good to me. Thanks @jhorstmann

**Review by sunchao — APPROVED** [review]
LGTM2

**Comment by alamb** [comment]
I merged this PR with `master` and fixed a rustdoc issue with e6c6c07dee https://github.com/apache/arrow-rs/runs/5756218032?check_suite_focus=true -- thanks @jhorstmann

**Comment by alamb** [comment]
I believe the clippy errors are due to rust 1.60 being released. I will create a new PR

**Pull request merged**